array.hpp

```diff
@@ -0,0 +1,282 @@
+        const_iterator cbegin()const;
+        const_iterator cend()const;
+        reverse_iterator rbegin();
+        const_reverse_iterator rbegin()const;
+        reverse_iterator rend();
+        const_reverse_iterator rend()const;
+        const_reverse_iterator crbegin()const;
+        const_reverse_iterator crend()const;
+
+
+    private:
+        T _arr[N];
+    };
+
+    template<typename T, std::size_t N>
+    constexpr  typename array<T,N>::const_reference array<T, N>::operator[](typename array::difference_type index) const noexcept {
+        return _arr[index];
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::reference array<T, N>::operator[](typename array::difference_type index) noexcept {
+        return _arr[index];
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::reference array<T, N>::at(array::difference_type index) {
+        if(index>=N){
+            throw std::out_of_range("Index of array's element must be less then it's size");
+        }
+        return _arr[index];
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::const_reference array<T, N>::at(array::difference_type index)const {
+        if(index>=N){
+            throw std::out_of_range("Index of array's element must be less then it's size");
+        }
+        return _arr[index];
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::pointer array<T, N>::data() noexcept {
+        return _arr;
+    }
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::const_pointer array<T, N>::data()const noexcept {
+        return _arr;
+    }
+
+    template<typename T, std::size_t N>
+    constexpr void array<T, N>::fill(const value_type & value) {
+        for (size_type i=0;i<N;++i){
+            _arr[i] = value;
+        }
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::reference array<T, N>::back()noexcept {
+        return _arr[N-1];
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::const_reference array<T, N>::back() const noexcept{
+        return _arr[N-1];
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::reference array<T, N>::front() noexcept{
+        return _arr[0];
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::const_reference array<T, N>::front() const noexcept{
+        return _arr[0];
+    }
+
+    template<typename T, std::size_t N>
+    constexpr bool array<T, N>::empty() const {
+        return N==0;
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::size_type array<T, N>::size() const {
+        return N;
+    }
+
+    template<typename T, std::size_t N>
+    constexpr typename array<T,N>::size_type array<T, N>::max_size() const {
+        return std::numeric_limits<size_type>::max();
```

Comment:
please note 

> Because each std::array<T, N> is a fixed-size container, the value returned by max_size equals N (which is also the value returned by size)